src/main/java/org/openmaptiles/layers/Water.java

```diff
@@ -121,11 +175,123 @@ public void process(Tables.OsmWaterPolygon element, FeatureCollector features) {
         .setAttr(Fields.INTERMITTENT, element.isIntermittent() ? 1 : 0)
         .setAttrWithMinzoom(Fields.BRUNNEL, Utils.brunnel(element.isBridge(), element.isTunnel()), 12)
         .setAttr(Fields.CLASS, clazz);
+
+      try {
+        var geom = element.source().worldGeometry();
+        try {
+          attemptNeLakeIdMapping(element, geom);
+        } catch (TopologyException e) {
+          try {
+            var fixedGeom = GeometryFixer.fix(geom);
+            attemptNeLakeIdMapping(element, fixedGeom);
+          } catch (TopologyException e2) {
+            throw new GeometryException("fix_omt_water_topology_error",
+              "error fixing polygon: " + e2 + "; original error: " + e);
+          }
+        }
+      } catch (GeometryException e) {
+        e.log(stats, "omt_water",
+          "Unable to process intersections for OSM feature " + element.source().id());
+      }
+    }
+  }
+
+  void attemptNeLakeIdMapping(Tables.OsmWaterPolygon element, Geometry geom) throws GeometryException {
+    // if OSM lake is too small for Z6 (e.g. area bellow ~4px) we assume there is no matching NE lake
+    if (geom.getArea() < OSM_ID_MATCH_AREA_LIMIT) {
+      return;
+    }
+
+    // match by name:
+    boolean match = false;
+    if (element.name() != null) {
+      for (var map : neLakeNameMaps.values()) {
+        var lakeInfo = map.get(element.name());
+        if (lakeInfo != null) {
+          match = true;
+          fillOsmIdIntoNeLake(element, geom, lakeInfo, true);
+        }
+      }
+    }
+    if (match) {
+      return;
+    }
+
+    // match by intersection:
+    var items = neLakeIndex.getIntersecting(geom);
+    for (var lakeInfo : items) {
+      fillOsmIdIntoNeLake(element, geom, lakeInfo, false);
+    }
+  }
+
+  /*
+   * When we match lakes with `neLakeIndexes` then `intersetsCheckNeeded` should be `false`,
+   * otherwise `true`, to make sure we DO check the intersection but to avoid checking it twice.
+   */
+  void fillOsmIdIntoNeLake(Tables.OsmWaterPolygon element, Geometry geom, LakeInfo lakeInfo,
+    boolean intersetsCheckNeeded) {
+    final Geometry neGeom = lakeInfo.geom;
+    if (intersetsCheckNeeded && !neGeom.intersects(geom)) {
+      return;
+    }
+    var intersection = neGeom.intersection(geom);
+
+    // Should match following in OpenMapTiles: Distinct on keeps just the first occurence -> order by 'area_ratio DESC'
+    // With a twist: NE geometry is always the same, hence we can make it a little bit faster by dropping "ratio"
+    // and compare only the intersection area: bigger area -> bigger ratio.
+    double area = intersection.getArea();
+    if (area > lakeInfo.area) {
+      lakeInfo.osmId = element.source().id();
+      lakeInfo.area = area;
+    }
```

Comment:
seems like there might be a risk here of multiple OSM threads updating the same lakeInfo instance at the same time, would probably be safer to wrap this in `synchronized (lakeInfo) { ... }`